### PR TITLE
feat: introduce frontend extenders

### PR DESCRIPTION
**Pull request by SychO9**
**Changes proposed in this pull request:**
Prior to this, frontend extenders never actually worked before, nor could any extension use them at all as they weren't exported in `compat` either. This PR does the following:
* Refactor extenders in core for them to actually work:
   - They are now in the `common/extenders` directory to avoid conflicts with the `extend.ts` module).
   - An `IExtender` interface has been introduced.
   - Th `Routes` and `PostTypes` extenders have been adapted to the new changes and the `Model` extender was removed for a separate PR.
* Make use of the `Routes` and `PostTypes` extenders in bundled extensions.
* The extenders can be accessed by exposing an `extend` module from the extension, which contains an array of extenders.

**Reviewers should focus on:**
* JS Tests are only failing because `check-typings` is running before new typings are built.
* The bundled extension changes were made here as an example, but I can extract them into another PR if wanted. The commits are separated though, so it should be easy to review.

**Necessity**

- [x] Has the problem that is being solved here been clearly explained?
- [x] If applicable, have various options for solving this problem been considered?
- [x] For core PRs, does this need to be in core, or could it be in an extension?
- [x] Are we willing to maintain this for years / potentially forever?

**Confirmed**

- [x] Frontend changes: tested on a local Flarum installation.
- [x] Backend changes: tests are green (run `composer test`).
- [ ] Core developer confirmed locally this works as intended.
- [x] Tests have been added, or are not appropriate here.

### CODE REVIEW

**Comment by luceos** [comment]
I don't feel prefixing interfaces with an `I` is the way to go. We are trying to mimic the ideology of Laravel mostly and Laravel explicitly moved away from `ISomething` and `SomethingInterface` to `Contracts\Something`. Is there a way to do this here as well? For instance naming the `IExtender` something like `Contract\Extends` or `Contract\Extensible`. This, in my experience with Laravel, seemed to be a more eloquent way of using interfaces honestly. What do you think?

**Comment by SychO9** [comment]
> I don't feel prefixing interfaces with an I is the way to go.

We are already using that convention with other frontend interfaces, so it would be inconsistent tbh.

> We are trying to mimic the ideology of Laravel mostly and Laravel explicitly moved away from ISomething and SomethingInterface to Contracts\Something.

That's on the backend side though, and we don't actually mimic Laravel, we explicitly follow the [PSR-2 coding style](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-2-coding-style-guide.md) which promotes `FooInterface`  `AbstractFoo` ...etc

I've changed the class name to `ExtenderInterface` though

File: framework/core/js/src/common/extenders/Routes.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+   * @param path The path of the route.
+   * @param component must extend `Page` component.
+   */
+  add(name: string, path: `/${string}`, component: any): Routes {
+    this.routes[name] = { path, component };
+
+    return this;
+  }
+
+  helper(name: string, callback: HelperRoute): Routes {
```
Comment (askvortsov1) [inline framework/core/js/src/common/extenders/Routes.ts:23]: Is the name "helper" consistent with the concrete implementations included in core's routing helpers? Can we somehow refer to that, even if the type doesn't adds any additional structure or safety, if only for documentation?

File: framework/core/js/src/common/extenders/Routes.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+    return this;
+  }
+
+  extend(app: Application, extension: ExtensionModuleInterface) {
+    Object.assign(app.routes, this.routes);
+
+    if (Object.keys(this.helpers).length) {
+      Object.assign(app.route, this.helpers);
+    }
```
Comment (askvortsov1) [inline framework/core/js/src/common/extenders/Routes.ts]: Why is this conditional? Assigning in an empty object should be a no-op, so why not do it regardless?

**Comment by askvortsov1** [comment]
> > I don't feel prefixing interfaces with an I is the way to go.
> 
> We are already using that convention with other frontend interfaces, so it would be inconsistent tbh.

I would prefer being consistent across the frontend. Different languages have different coding styles, I prefer to stick to convention in each as opposed to forcing an odd middle ground.

File: framework/core/js/src/common/extenders/Routes.ts (hunk 1 above)
Comment (SychO9) [reply framework/core/js/src/common/extenders/Routes.ts:23]: It's referred to as `routeHelpers` in core, so I think `helper` makes sense.

https://github.com/flarum/framework/blob/a595665bfbf7820c58a3911df154ced435b6fbfd/framework/core/js/src/forum/routes.ts#L41-L51

File: framework/core/js/src/common/extenders/Routes.ts (hunk 2 above)
Comment (SychO9) [reply framework/core/js/src/common/extenders/Routes.ts]: I honestly don't recall why I specifically did this here and not elsewhere, it is unnecessary.

**Review by askvortsov1 — APPROVED** [review]
Great step towards a more consistent, declarative API. Sorry for the review delay!

**Pull request merged**